src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs

```diff
@@ -150,147 +155,232 @@ internal partial class Session
 
-                var result = model.WithFilteredItems(allFilteredItems.ToImmutableArray())
-                                  .WithSelectedItem(selectedItem)
+            private Model HandleNormalFiltering(
+                Model model, CompletionFilterReason filterReason,
+                ITextSnapshot textSnapshot, Document document,
+                CompletionHelper helper, ImmutableArray<string> recentItems,
+                string filterText,
+                List<FilterResult> filterResults)
+            {
+                // Not deletion.  Defer to the language to decide which item it thinks best
+                // matches the text typed so far.
+
+                // Ask the language to determine which of the *matched* items it wants to select.
+                var service = this.Controller.GetCompletionService();
+
+                var matchingCompletionItems = filterResults.Where(r => r.MatchedFilterText)
+                                                           .Select(t => t.PresentationItem.Item)
+                                                           .AsImmutable();
+                var chosenItems = service.ChooseBestItems(document, matchingCompletionItems, filterText);
+
+                // Of the items the service returned, pick the one most recently committed
+                var bestCompletionItem = GetBestCompletionItemBasedOnMRU(chosenItems, recentItems);
+
+                // If we don't have a best completion item yet, then pick the first item from the list.
+                var bestOrFirstCompletionItem = bestCompletionItem ?? filterResults.First().PresentationItem.Item;
+                var bestOrFirstPresentationItem = filterResults.Where(
+                    r => r.PresentationItem.Item == bestOrFirstCompletionItem).First().PresentationItem;
+
+                var hardSelection = IsHardSelection(
+                    model, bestOrFirstPresentationItem, textSnapshot, helper, filterReason);
+
+                // Determine if we should consider this item 'unique' or not.  A unique item
+                // will be automatically committed if the user hits the 'invoke completion' 
+                // without bringing up the completion list.  An item is unique if it was the
+                // only item to match the text typed so far, and there was at least some text
+                // typed.  i.e.  if we have "Console.$$" we don't want to commit something
+                // like "WriteLine" since no filter text has actually been provided.  HOwever,
+                // if "Console.WriteL$$" is typed, then we do want "WriteLine" to be committed.
+                var matchingItemCount = filterResults.Where(t => t.MatchedFilterText).Count();
+                var isUnique = bestCompletionItem != null &&
+                    matchingItemCount == 1 &&
+                    filterText.Length > 0;
+
+                var result = model.WithFilteredItems(filterResults.Select(r => r.PresentationItem).AsImmutable())
```

Comment:
So "matches filter text" is not equivalent to "included in completion list".  

For example, if only one character was written (like 'x'), then lots of things won't match it, but they'll still be included.

We keep around the things like 'matches filter text' so we can make other determinations (i.e. what to select, and if it is unique).  However, we still want to display all the items we decided to filter down to (even if they didn't match the filter text).